pyquil/quilbase.py

```diff
@@ -244,9 +145,22 @@ def __init__(self, name, matrix):
         self.name = name
         self.matrix = np.asarray(matrix)
 
-        is_unitary = np.allclose(np.eye(rows), self.matrix.dot(self.matrix.T.conj()))
-        if not is_unitary:
-            raise ValueError("Matrix must be unitary.")
+        if parameters:
+            if not isinstance(parameters, list):
+                raise TypeError("Paramaters must be a list")
+
+            expressions = [elem for row in self.matrix for elem in row if isinstance(elem, Expression)]
+            used_params = {param for exp in expressions for param in _contained_parameters(exp)}
+
+            if set(parameters) != used_params:
+                raise ValueError("Parameters list does not match parameters actually used in gate matrix:\n"
+                                 "Parameters in argument: {}, Parameters in matrix: {}".format(parameters, used_params))
+        else:
+            is_unitary = np.allclose(np.eye(rows), self.matrix.dot(self.matrix.T.conj()))
```

Comment:
Not necessary, but we could verify unitarity for a few randomly sampled values of the parameters as well